ashura/engine/animation.h

```diff
@@ -2,25 +2,1010 @@
+                     }
+                     return lerp(start, end, factor);
+                   }};
+}
+
+// Spring-based elastic easing
+// Based on simple harmonic motion with damping
+// Reference: https://www.ryanjuckett.com/damped-springs/
+// - mass: Oscillator mass (default: 1.0)
+// - stiffness: Spring constant (default: 20.0)
+// - damping: Damping coefficient (default: 10.0f)
+//
+// The default behavior is a tight spring effect, tune the parameters to give
+// a desired effect.
+template <Animatable T>
+constexpr Easing<T> spring(f32 mass = 1.0f, f32 stiffness = 20.0f,
+                           f32 damping = 10.0f)
+{
+  return Easing<T>{
+      .params = {0.0f, mass, stiffness, damping},
+      .easing = [](const T &start, const T &end, f32 *params) {
+        f32 t         = params[0];        // current time
+        f32 mass      = params[1];        // Oscillator mass
+        f32 stiffness = params[2];        // Spring constant
+        f32 damping   = params[3];        // Damping coefficient
+
+        // Calculate critical damping factors
+        f32 omega0           = std::sqrt(stiffness / mass);
+        f32 critical_damping = 2.0f * std::sqrt(mass * stiffness);
+        f32 zeta = damping / critical_damping;        // damping ratio
+
+        if (zeta < 1.0f)
+        {        // Underdamped
+          f32  omega_d = omega0 * std::sqrt(1.0f - zeta * zeta);
+          auto factor =
+              1.0f - std::exp(-zeta * omega0 * t) *
+                         (std::cos(omega_d * t) +
+                          (zeta * omega0 / omega_d) * std::sin(omega_d * t));
+          return lerp(start, end, factor);
+        }
+        // Overdamped or critically damped
+        f32  alpha = -zeta * omega0;
+        f32  beta  = omega0 * std::sqrt(zeta * zeta - 1.0f);
+        auto factor =
+            1.0f -
+            (std::exp(alpha * t) *
+             (std::cosh(beta * t) + (alpha / beta) * std::sinh(beta * t)));
+
+        return lerp(start, end, factor);
+      }};
+}
+
+}        // namespace anim
+
+template <Animatable T>
+struct Easing
+{
+  typedef T (*Function)(const T &start, const T &end, f32 *params);
+  f32      params[4] = {};
+  Function easing    = nullptr;
+
+  T evaluate(const T &start, const T &end, f32 t) const
+  {
+    if (easing)
+    {
+      f32 temp_params[4] = {t};
+      for (int i = 1; i < 4; ++i)
+      {
+        temp_params[i] = params[i];
+      }
+      return easing(start, end, temp_params);
+    }
+
+    // Default to linear interpolation
+    if constexpr (std::is_same_v<T, Vec2> || std::is_same_v<T, Vec3> ||
+                  std::is_same_v<T, Vec4> || std::is_same_v<T, Vec4U8>)
+    {
+      return anim::lerp(start, end, t);        // Uses the splat version
+    }
+    else
+    {
+      return anim::lerp(start, end, T(t));        // Convert t to type T
+    }
+  }
+
+  // TODO: More Easing functions , see https://github.com/EmmanuelOga/easing
+  // - outInQuad
```

Comment:
I don't really think these are useful, unless there's an explicit request for it

Reply:
Yeah, though it would be convenient for devs. Most people wouldn't want to write custom easing functions. To be honest, this is the API side of most animation libraries 😄